v3/src/components/case-table/use-rows.ts

```diff
@@ -222,18 +224,96 @@ export const useRows = () => {
+    const casesToUpdate: ICase[] = []
+    const casesToCreate: ICaseCreation[] = []
+    caseValues.forEach(aCase => {
+      if (aCase.__id__ === kInputRowKey) {
+        const { __id__, ...others } = aCase
+
+        // Do not add a new item if no actual values are specified.
+        // This can happen when a user starts editing a cell in the input row, but then natigates away with
+        // the cell left blank
+        if (!Object.values(others).some(value => !!value)) return
+
+        // Find values inherited from parent case
+        const prevRowIndex = inputRowIndex != null && inputRowIndex !== -1
+          ? inputRowIndex > 0
+            ? inputRowIndex - 1
+            : 1
+          : _rows.length - 2
+        const prevRowId = _rows[prevRowIndex].__id__
+        const prevRow = collectionTableModel?.rowCache.get(prevRowId)
+        const parentId = prevRow?.[symParent]
+        const parentValues = data?.getParentValues(parentId ?? "") ?? {}
+        
+        casesToCreate.push({ ...others, ...parentValues })
+      } else {
+        casesToUpdate.push(aCase)
+      }
+    })
+
+    const creatingCases = casesToCreate.length > 0
+    const undoStringKey = creatingCases ? "DG.Undo.caseTable.createNewCase" : "DG.Undo.caseTable.editCellValue"
+    const redoStringKey = creatingCases ? "DG.Redo.caseTable.createNewCase" : "DG.Redo.caseTable.editCellValue"
+
+    // We track case ids between updates and additions so we can make proper notifications afterwards
+    let oldCaseIds = new Set(collection?.caseIds ?? [])
+    let updatedCaseIds: string[] = []
+    const newCaseIds: string[] = []
     data?.applyModelChange(
-      () => data.setCaseValues(caseValues),
+      () => {
+        // Update existing cases
+        if (casesToUpdate.length > 0) {
+          data.setCaseValues(casesToUpdate)
+          if (collection?.id === data.childCollection.id) {
+            // The child collection's case ids are persistent, so we can just use the casesToUpdate to
+            // determine which case ids to use in the updateCasesNotification
+            updatedCaseIds = casesToUpdate.map(aCase => aCase.__id__)
+          } else {
+            // Other collections have cases whose ids change when values change, so we have to check which case ids
+            // were not present before updating to determine which case ids to use in the updateCasesNotification
+            collection?.caseIds.forEach(caseId => {
+              if (!oldCaseIds.has(caseId)) updatedCaseIds.push(caseId)
+            })
+          }
+          oldCaseIds = new Set(collection?.caseIds ?? [])
+        }
+
+        // Create new cases
+        if (creatingCases) {
+          const options: IAddCasesOptions = {}
+          if (collectionTableModel?.inputRowIndex != null && collectionTableModel.inputRowIndex >= 0) {
+            options.before = collection?.caseIds[collectionTableModel.inputRowIndex]
+            collectionTableModel.inputRowIndex += 1
```

Comment:
Should use an `@action` here rather than updating the model directly:
```suggestion
            collectionTableModel.setInputRowIndex(collectionTableModel.inputRowIndex + 1)
```